packages/flutter_tools/lib/src/extension/device.dart

```diff
@@ -0,0 +1,311 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// Tool extensions for interfacing with a flutter capable device.
+library device;
+
+import 'package:meta/meta.dart';
+
+import 'extension.dart';
+
+/// A description of the kind of workflow the device supports.
+///
+/// This information provides hints to IDEs and edtiors that
+/// integrate with the flutter tool.
+class Category implements Serializable {
+  const Category._(this._value);
+
+  /// Create a new [Category] from a json object.
+  ///
+  /// Throws [ArgumentError] if provided with an invalid value.
+  factory Category.fromJson(int value) {
+    switch (value) {
+      case 0:
+        return web;
+      case 1:
+        return desktop;
+      case 2:
+        return mobile;
+    }
+    throw ArgumentError.value(value);
+  }
+
+  /// A category of devices with a web workflow.
+  static const Category web = Category._(0);
+
+  /// A category of devices with a desktop workflow.
+  static const Category desktop = Category._(1);
+
+  /// A category of devices with a mobile workflow.
+  static const Category mobile = Category._(2);
+
+  final int _value;
+
+  @override
+  Object toJson() {
+    return _value;
+  }
+}
+
+/// A description of the platform Rlutter will run on.
+class TargetPlatform implements Serializable {
+  const TargetPlatform._(this._value);
+
+  /// Create a new [TargetPlatform] from a json encoded value.
+  ///
+  /// Throws [ArgumentError] if an invalid value is provided.
+  factory TargetPlatform.fromJson(int value) {
+    switch (value){
+      case 0:
+        return android;
+      case 1:
+        return iOS;
+      case 2:
+        return windows;
+      case 3:
+        return linux;
+      case 4:
+        return macOS;
+      case 5:
+        return fuchsia;
+      case 6:
+        return web;
+    }
```

Comment:
TargetPlatform is a bit complicated in that it overloads a bunch of difference concepts, like what AOT artifacts and what platform folder we're using.  In theory this is solved by making flutter assemble extensible enough and exposing it via the extension API.

I don't have a good solution for now though, perhaps besides documenting that this should be deprecated and removed before 1.0, or else otherwise replaced with a String like in Category.